CRM/Contribute/Page/UserDashboard.php

```diff
@@ -18,112 +21,84 @@ class CRM_Contribute_Page_UserDashboard extends CRM_Contact_Page_View_UserDashBo
 
   /**
    * called when action is browse.
+   *
+   * @throws \CRM_Core_Exception
    */
-  public function listContribution() {
-    $rows = civicrm_api3('Contribution', 'get', [
-      'options' => [
-        'limit' => 12,
-        'sort' => 'receive_date DESC',
-      ],
-      'sequential' => 1,
-      'contact_id' => $this->_contactId,
-      'return' => [
-        'total_amount',
-        'contribution_recur_id',
-        'financial_type',
-        'receive_date',
-        'receipt_date',
-        'contribution_status',
-        'currency',
-        'amount_level',
-        'contact_id,',
-        'contribution_source',
-      ],
-    ])['values'];
-
-    // We want oldest first, just among the most recent contributions
-    $rows = array_reverse($rows);
+  public function listContribution(): void {
+    $contributions = $this->getContributions();
 
-    foreach ($rows as $index => &$row) {
+    foreach ($contributions as &$row) {
       // This is required for tpl logic. We should move away from hard-code this to adding an array of actions to the row
       // which the tpl can iterate through - this should allow us to cope with competing attempts to add new buttons
       // and allow extensions to assign new ones through the pageRun hook
-      $row['balance_amount'] = CRM_Contribute_BAO_Contribution::getContributionBalance($row['contribution_id']);
-      $contributionStatus = CRM_Core_PseudoConstant::getName('CRM_Contribute_BAO_Contribution', 'contribution_status_id', $row['contribution_status_id']);
-
-      if (in_array($contributionStatus, ['Pending', 'Partially paid'])) {
+      // We could check for balance_amount > 0 here? It feels more correct but this seems to be working.
+      if (in_array($row['contribution_status_id:name'], ['Pending', 'Partially paid'], TRUE)) {
         $row['buttons']['pay'] = [
           'class' => 'button',
           'label' => ts('Pay Now'),
           'url' => CRM_Utils_System::url('civicrm/contribute/transact', [
             'reset' => 1,
             'id' => Civi::settings()->get('default_invoice_page'),
-            'ccid' => $row['contribution_id'],
+            'ccid' => $row['id'],
             'cs' => $this->getUserChecksum(),
             'cid' => $row['contact_id'],
           ]),
         ];
       }
     }
+    unset($row);
```

Comment:
@eileenmcnaughton why is this here?

Reply:
it's just for clarity when you have a variable bleed out of a loop & it is a reference there is a bit of a risk of doing something odd with it